forem/forem
# Update storybook autocomplete component

## What type of PR is this? (check all applicable)


## Description

This PR reworks the MentionAutocomplete component initially created in #12831. After some more careful testing with the previous implementation we decided the Reach UI Combobox needs to fully take over the target text area. This allows us to cover a greater variety of triggers (e.g. deleting back to a previous word with `@mention`) in a more dependable way.

The `MentionAutocompleteTextArea` component now works by fully replacing the element passed to in props (copying all styles and attributes), and initialises the search behaviour when the given word (based on cursor position) begins with an `@`.

## Related Tickets & Documents

[Create mention autocomplete](https://github.com/forem/rfcs/pull/16)

## QA Instructions, Screenshots, Recordings

This can be tested in Storybook using the netlify deploy preview linked below

- Type text in the text area - you should notice no layout shifts as it is replaced with the autocomplete text area
- Experiment with adding `@mentions` in various places in the text content
- You should be able to select an option either by using down/up arrow keys and pressing enter, or by clicking an item
- After selection, the cursor should return to the end of the inserted mention
- You should also be able to just keep typing and ignore any suggestions, with the dropdown disappearing when you space away from the word
- Pressing the Escape key should hide the popup (but once you start typing again it should re-appear)
- Clicking outside of the dropdown should also close it

https://user-images.githubusercontent.com/20773163/110931920-b8f32000-8322-11eb-93c8-858066ab081d.mp4


### UI accessibility concerns?

This refactor means that the whole textarea is now assigned the role "combobox". Screen reader users are immediately informed that a list of options will become associated with the input. An `aria-live` has been added to announce to screen reader users when the `@mention` search has started populating the list of options.

It's important to note that axe flags an error for this component, as according to the [WAI ARIA guidelines](https://www.w3.org/TR/wai-aria-practices/#combobox) a combobox role should not apply to a `textarea` but a single line input. We have had a lot of discussions about how best to handle the screen reader use-case, and the implementation in this PR is very similar to the one GitHub is using for mentions. It makes sure that screen reader users can take advantage of the autocomplete functionality, despite it taking place in a textarea.

I also don't believe this implementation will degrade the current editor/comment experience for screen reader users (when added there), since in all other respects the textarea continues to operate as before. 

In short, I _think_ continuing in this direction is the best a11y experience we can offer to the best of our knowledge right now, but we should look in future to get this audited/tested by real users of assistive technology and be open to any feedback we receive.

## Added tests?

- [X] Yes (although much of the functionality will be covered by Cypress tests when implemented in the app)

## [Forem core team only] How will this change be communicated?

- [X] This change does not need to be communicated, and this is why not: This only affects the Storybook component and changes will be communicated when the component is actually implemented in the app


## [optional] What gif best describes this PR or how it makes you feel?

![Baga Chipz says much betta](https://i.giphy.com/media/jqMXe0AgWsHZYA5mjd/giphy.webp)
